src/Raven.Server/ServerWide/Context/FilesOperationContext.cs

```diff
@@ -0,0 +1,94 @@
+    {
+        private readonly FileSystem _fileSystem;
+
+        public short TransactionMarkerOffset;
+
+        private short _currentTxMarker;
+
+        public short GetTransactionMarker()
+        {
+            if (Transaction != null && Transaction.Disposed == false && Transaction.InnerTransaction.LowLevelTransaction.Flags != TransactionFlags.ReadWrite)
+                MustHaveWriteTransactionOpened();
+
+            var value = (short)(_currentTxMarker + TransactionMarkerOffset);
+            if (value <= 0)
+            {
+                switch (value)
+                {
+                    case short.MaxValue:
+                        return 1;
+                    case 0:
+                        return 2;
+                    default:
+                        return (short) -value;
+                }
+            }
+            return value;
+        }
+
+        private static void MustHaveWriteTransactionOpened()
+        {
+            throw new InvalidOperationException("Write transaction must be opened");
+        }
+
+        public static FilesOperationContext ShortTermSingleUse(FileSystem fileSystem)
+        {
+            var shortTermSingleUse = new FilesOperationContext(fileSystem, 4096, 1024);
+            return shortTermSingleUse;
+        }
+
+        public FilesOperationContext(FileSystem fileSystem, int initialSize, int longLivedSize) :
+            base(initialSize, longLivedSize)
+        {
+            _fileSystem = fileSystem;
+        }
+
+        protected override FilesTransaction CreateReadTransaction()
+        {
+            return new FilesTransaction(this, _fileSystem.FilesStorage.Environment.ReadTransaction(PersistentContext, Allocator));
+        }
+
+        protected override FilesTransaction CreateWriteTransaction()
+        {
+            var tx = new FilesTransaction(this, _fileSystem.FilesStorage.Environment.WriteTransaction(PersistentContext, Allocator));
+
+            _currentTxMarker = (short) tx.InnerTransaction.LowLevelTransaction.Id;
+
+            var options = _fileSystem.FilesStorage.Environment.Options;
+
+            if ((options.TransactionsMode == TransactionsMode.Lazy || options.TransactionsMode == TransactionsMode.Danger) &&
+                options.NonSafeTransactionExpiration != null && options.NonSafeTransactionExpiration < DateTime.Now)
+            {
+                options.TransactionsMode = TransactionsMode.Safe;
+            }
+
+            tx.InnerTransaction.LowLevelTransaction.IsLazyTransaction = 
+                options.TransactionsMode == TransactionsMode.Lazy;
+            // IsLazyTransaction can be overriden later by a specific feature like bulk insert
+
+            return tx;
+        }
+
+        public StorageEnvironment Environment => _fileSystem.FilesStorage.Environment;
+
+        public FileSystem FileSystem => _fileSystem;
+
+        public bool ShouldRenewTransactionsToAllowFlushing()
```

Comment:
It is OK to have the duplication now, as I just start to see what is needed by the FileSystem. Once it will be solid we will merge staff.